#574 Port org.biojava.bio.program.abi.ABITracer

MaxGreil opened the pull request:
#574 
Added 3 new classes: `ABITrace.java`, `ABITracerTest.java` and `ABITracerCompoundSet.java`.
The test file `3730.ab1` is from https://github.com/biopython/biopython/blob/master/Tests/Abi/3730.ab1.
This file was mentioned in #535.
The test data for file `3730.abi1` is from https://github.com/biopython/biopython/blob/master/Tests/Abi/test_data.

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+    //the next three lines are the important persistent data
+    private String sequence;
+    private int A[], G[], C[], T[], Basecalls[], Qcalls[];
+    private int TraceLength, SeqLength;
```
Variables should start with lower case as per normal java conventions.

I know this was like that in legacy project, but it'd be a great opportunity to fix the style while porting it :)

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+import org.biojava.nbio.core.exceptions.CompoundNotFoundException;
+
+/**
+ * Title: ABITrace<br><br>
```
Paragraph separator should be `<P>` instead of `<br>`, see https://stackoverflow.com/questions/5260368/which-tag-should-be-used-as-paragraph-separator-in-javadoc?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+    //of crap pre-pended to them. This constant
+    //allows ABITrace to handle that in a way that
+    //is invisible to the user.
+    private static int AbsIndexBase = 26; //The file location of the Index pointer
```
Constants should be `static final`

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+     * @throws IOException if there is a problem reading from the URL.
+     * @throws IllegalArgumentException if the URL does not contain a valid ABI file.
+     */
+    public ABITrace( URL ABIFile ) throws IOException
```
This and the method above are very similar. It'd be nice to extract the common part to a separate method to be shared by both.

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+        try {
+            DNASequenceCreator creator = new DNASequenceCreator(ABITracerCompoundSet.getABITracerCompoundSet());
+            return creator.getSequence(sequence, 0);
+        } catch (Exception e) {
```
No need to catch, just let the `CompoundNotFoundException` go through.

Catching a general `Exception` here could cause undesired side effects in the error handling.

If you think the `CompoundNotFoundException` can't happen (other than because of a bug), then it's valid to catch it (but only that exception), log a warning and not throw it.

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+    /**
+     * Returns the length of the sequence (number of bases) in this trace.
+     */
+    public int getSequenceLength() {
```
The `@return` tags are missing in a few javadocs

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+import java.io.InputStream;
+
+import org.biojava.nbio.core.sequence.compound.ABITracerCompoundSet;
+import org.biojava.nbio.core.sequence.compound.DNACompoundSet;
```
Unused import

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+     * Returns the scaling factor necessary to allow all of the traces to fit vertically
+     * into the specified space.
+     *
+     * @param - the required height in pixels.
```
Missing param name

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,568 @@
+            getSubArray(RecNameArray, (IndexBase + (record * 28)));
+            RecName = new String(RecNameArray);
+            if (RecName.equals("FWO_"))
+                FWO = IndexBase + (record * 28) + 20;
```
Any idea what this 2 "magic" numbers mean? It'd be nice to extract them to constants with some documentation as to what they mean.

MaxGreil replied:
I don't know what these 2 numbers stand for. What would be the best names to call them?

josemduarte replied:
I'm not familiar with the format either. They seem some kind of offset, but I can't really tell.

FYI the format is described http://www6.appliedbiosystems.com/support/software_community/ABIF_File_Format.pdf

Let's leave it as is then... unless there's other suggestions, @el1jaC ?

josemduarte commented on biojava-core/src/test/java/org/biojava/nbio/core/sequence/io/ABITracerTest.java:
```diff
@@ -0,0 +1,97 @@
+ */
+public class ABITracerTest {
+
+    private String filePath = System.getProperty("user.dir") + "/src/test/resources/3730.ab1";
```
This is not going to work when running as a jar. See how to do it here: https://stackoverflow.com/questions/3891375/how-to-read-a-text-file-resource-into-java-unit-test?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa

MaxGreil replied:
I have a question regarding this. 
Is it possible to add the dependency for org.apache.commons.io.IOUtils? 
Its function IOUtils can be used to safely convert the InputStream to a byte[] and this can then be used for the Constructor ABITracer (byte[] ABIFileData).

MaxGreil commented:
@josemduarte 
* Refactored constructors `ABITrace(File ABIFile)` and `ABITrace( URL ABIFile )` with new shared method `private void ABITraceInit(BufferedInputStream bis)`. 
* Corrected all other requests. 
* Left request for method `private void setIndex()`.

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,597 @@
+    //allows ABITrace to handle that in a way that
+    //is invisible to the user.
+    private static final int absIndexBase = 26; //The file location of the Index pointer
+    private int IndexBase, PLOC, PCON;
```
IndexBase should be lower case, plus the variable should be local to the method where it is used.

josemduarte commented on biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/ABITrace.java:
```diff
@@ -0,0 +1,597 @@
+    //This is the actual file data.
+    private byte[] traceData;
+
+    private int maximum = 0;
```
Doesn't look that this is being used. I seems to be a caching variable but not used properly. I'd remove it.

MaxGreil commented:
Fixed minor changes.

MaxGreil commented:
@josemduarte 
Thank you. Is the PR now ready for merge with the master branch?

josemduarte commented:
I'd like to leave it open for a couple of days in case there are any other comments.

el1jaC commented:
@josemduarte sorry for the late, I didn't see the tag!

Btw, I'm very happy that classes are ported to new Biojava.

josemduarte commented:
> When the pull request will be merged, can I add support in ABITrace class to getImageReverted, to manage RC sequences?

Absolutely, more than welcomed!

josemduarte commented:
@el1jaC please go ahead and submit the new pull request when you are ready. This one is merged.

el1jaC commented:
Thank you  @josemduarte .

I'll do this in the next days!
